File: app/src/main/java/jp/sourceforge/gokigen/memoma/extension/ExtensionActivity.java
package jp.sourceforge.gokigen.memoma.extension;

import android.content.Intent;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import jp.sourceforge.gokigen.memoma.R;

/**
 *
 */
public class ExtensionActivity extends AppCompatActivity
{
	// 起動コード
    public static final String MEMOMA_EXTENSION_LAUNCH_ACTIVITY = "jp.sfjp.gokigen.memoma.extension.activity";

    // データ識別子(表示中データの保存ファイルへのフルパス)
    public static final String MEMOMA_EXTENSION_DATA_FULLPATH = "jp.sfjp.gokigen.memoma.extension.data.fullpath";
    public static final String MEMOMA_EXTENSION_DATA_TITLE = "jp.sfjp.gokigen.memoma.extension.data.title";

    private ExtensionActivityListener listener = null;
    
	/** Called when the activity is first created. */
    @Override
    public void onCreate(Bundle savedInstanceState) 
    {
          super.onCreate(savedInstanceState);

          try
          {
              // リスナクラスを生成する
              listener = new ExtensionActivityListener(this);

              // レイアウトを設定する
              setContentView(R.layout.extensionview);

              // リスナクラスの準備
              listener.prepareExtraDatas(getIntent());
              listener.prepareListener();
          }
          catch (Exception e)
          {
              e.printStackTrace();
          }
    }

    /**
     *  メニューの生成
     */
    @Override
    public boolean onCreateOptionsMenu(Menu menu)
    {
        try
        {
            menu = listener.onCreateOptionsMenu(menu);
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
        return (super.onCreateOptionsMenu(menu));
    }

    /**
     *  メニューアイテムの選択
     */
    @Override
    public boolean onOptionsItemSelected(MenuItem item)
    {
        return (listener.onOptionsItemSelected(item));
    }
    
    /**
     *  メニュー表示前の処理
     */
    @Override
    public boolean onPrepareOptionsMenu(Menu menu)
    {
        try
        {
            listener.onPrepareOptionsMenu(menu);
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
        return (super.onPrepareOptionsMenu(menu));
    }

    /**
     *  画面が裏に回ったときの処理
     */
    @Override
    public void onPause()
    {
        super.onPause();
        try
        {
            // 動作を止めるようイベント処理クラスに指示する
        	listener.shutdown();        	
        }
        catch (Exception ex)
        {
        	// 何もしない
        }
    }
    
    /**
     *  画面が表に出てきたときの処理
     */
    @Override
    public void onResume()
    {
        super.onResume();
        try
        {
        	// 動作準備するようイベント処理クラスに指示する
        	listener.prepareToStart();
        }
        catch (Exception ex)
        {
            // なにもしない
            ex.printStackTrace();
        }
    }

    /**
     *   終了時の処理
     * 
     */
    @Override
    protected void onDestroy()
    {
        try
        {
            listener.finishListener();
            super.onDestroy();
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
    }

    /**
     * 
     */
    @Override
    protected void onStart()
    {
        super.onStart();
    }

    /**
     * 
     */
    @Override
    protected void onStop()
    {
        super.onStop();
    }

    /**
     * 
     */
    @Override
    protected void onSaveInstanceState(@NonNull Bundle outState)
    {
        super.onSaveInstanceState(outState);
    }

    /**
     * 
     */
    @Override
    protected void onRestoreInstanceState(Bundle savedInstanceState)
    {
        super.onRestoreInstanceState(savedInstanceState);
    }

    /**
     *  子画面から応答をもらったときの処理
     */
    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data)
    {
        super.onActivityResult(requestCode, resultCode, data);
        try
        {
            // 子画面からもらった情報の応答処理をイベント処理クラスに依頼する
            listener.onActivityResult(requestCode, resultCode, data);
        }
        catch (Exception ex)
        {
            // 例外が発生したときには、何もしない。
            ex.printStackTrace();
        }
    }    
}

File: app/src/main/java/jp/sourceforge/gokigen/memoma/extension/ExtensionActivity.kt
package jp.sourceforge.gokigen.memoma.extension

import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment

class ExtensionActivity : Fragment()
{
    private var listener: ExtensionActivityListener? = null
    init {
        val activity = this.activity as AppCompatActivity
        listener = ExtensionActivityListener(activity)
    }

    override fun onCreate(savedInstanceState: Bundle?)
    {
        super.onCreate(savedInstanceState)
        try
        {
            Log.v(TAG, "ExtensionActivity::onCreate()")

            //// レイアウトを設定する
            //setContentView(R.layout.extensionview)

            // リスナクラスの準備
            val intent = this.activity?.intent
            if (intent != null)
            {
                listener?.prepareExtraDatas(intent)
            }
            listener?.prepareListener()

            // メニューがあるよ
            setHasOptionsMenu(true)
        }
        catch (e: Exception)
        {
            e.printStackTrace()
        }
    }

    /**
     * メニューの生成
     */
    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater)
    {
        try
        {
            listener?.onCreateOptionsMenu(menu)
        }
        catch (e: Exception)
        {
            e.printStackTrace()
        }
        return (super.onCreateOptionsMenu(menu, inflater))
    }

    /**
     * メニューアイテムの選択
     */
    override fun onOptionsItemSelected(item: MenuItem): Boolean
    {
        if (listener != null)
        {
            return (listener?.onOptionsItemSelected(item)?: false)
        }
        return (false)
    }

    /**
     * メニュー表示前の処理
     */
    override fun onPrepareOptionsMenu(menu: Menu)
    {
        try
        {
            super.onPrepareOptionsMenu(menu)
            listener?.onPrepareOptionsMenu(menu)
        }
        catch (e: Exception)
        {
            e.printStackTrace()
        }
    }

    /**
     * 画面が裏に回ったときの処理
     */
    public override fun onPause()
    {
        try
        {
            super.onPause()

            // 動作を止めるようイベント処理クラスに指示する
            listener?.shutdown()
        }
        catch (ex: Exception)
        {
            ex.printStackTrace()
        }
    }

    /**
     * 画面が表に出てきたときの処理
     */
    public override fun onResume()
    {
        try
        {
            super.onResume()
            listener?.prepareToStart()
        }
        catch (ex: Exception)
        {
            ex.printStackTrace()
        }
    }

    /**
     * 終了時の処理
     *
     */
    override fun onDestroy()
    {
        try
        {
            //listener?.finishListener()
            super.onDestroy()
        }
        catch (e: Exception)
        {
            e.printStackTrace()
        }
    }

    companion object {
        private val TAG = ExtensionActivity::class.java.simpleName

        // 起動コード
        const val MEMOMA_EXTENSION_LAUNCH_ACTIVITY: String = "jp.sfjp.gokigen.memoma.extension.activity"

        // データ識別子(表示中データの保存ファイルへのフルパス)
        const val MEMOMA_EXTENSION_DATA_FULLPATH: String = "jp.sfjp.gokigen.memoma.extension.data.fullpath"
        const val MEMOMA_EXTENSION_DATA_TITLE: String = "jp.sfjp.gokigen.memoma.extension.data.title"
    }
}